include/boost/geometry/algorithms/detail/overlay/split_rings.hpp

```diff
@@ -0,0 +1,433 @@
+
+
+            if (op1.seg_id.segment_index != op2.seg_id.segment_index)
+            {
+                return op1.seg_id.segment_index < op2.seg_id.segment_index;
+            }
+            return op1.fraction < op2.fraction;
+        }
+    };
+
+    template <typename InterruptPolicy>
+    static inline void get_self_turns(Ring const& ring,
+                                      turns_container_type& turns,
+                                      RobustPolicy const& robust_policy,
+                                      InterruptPolicy const& policy)
+    {  
+        geometry::self_turns
+            <
+                get_turn_info<assign_null_policy>
+            >(ring, robust_policy, turns, policy);
+    }
+
+    static inline void get_self_turns(Ring const& ring,
+                                      turns_container_type& turns,
+                                      RobustPolicy const& robust_policy)
+    {
+        get_self_turns(ring, turns, robust_policy, no_interrupt_policy());
+    }
+
+    template <typename MUU_Turns, typename RingOut>
+    static inline void insert_muu_turns(Ring const& ring,
+                                        MUU_Turns const& muu_turns,
+                                        RingOut& ring_out)
+    {
+        typedef typename boost::range_iterator
+            <
+                MUU_Turns const
+            >::type iterator_type;
+
+        typename boost::range_size<Ring>::type point_index = 0;
+        for (iterator_type it = muu_turns.begin(); it != muu_turns.end(); ++it)
+        {
+            signed_size_type segment_index
+                = get_correct_op(*it).seg_id.segment_index;
+
+            while (point_index <= segment_index)
+            {
+                ring_out.push_back(ring[point_index]);
+                ++point_index;
+            }
+
+            ring_out.push_back(it->point);
+        }
+
+        while (point_index < ring.size())
+        {
+            ring_out.push_back(ring[point_index]);
+            ++point_index; 
+        }
+    }
+
+    template <typename RingIn, typename Collection>
+    static inline void copy_to_collection(RingIn const& ring,
+                                          Collection& collection)
+    {
+        typedef typename boost::range_value<Collection>::type ring_out_type;
+
+        ring_out_type tmp;
+        for (typename boost::range_iterator<RingIn const>::type
+                 it = ring.begin();
+             it != ring.end();
+             ++it)
+        {
+            geometry::traits::push_back<ring_out_type>::apply(tmp, *it);
+        }
+        collection.push_back(tmp);
+    }
+
+    template <closure_selector Closure, typename RingType, typename Collection>
+    static inline void split_one_ring(RingType& ring, Collection& collection)
+    {
+        // create and initialize a stack with the input ring
+        std::stack<RingType> stack;
+        stack.push(RingType());
+        stack.top().swap(ring);
```

Comment:
Is this C++03 move emulation? :)
Is `RingType` adapted to Boost.Geometry Ring concept or `std::list<>`? I'm asking because Rings aren't required to implement `swap()`. So it seems that this code requires STL container. If the second part of my question is true then maybe it'd be a good idea to simply take `std::list` as parameter?

To make it even more clear, you could implement `move_on_top()` or `move_back()`. Similar algorithm is used below.